csrank/objectranking/lambdamart.py

```diff
@@ -0,0 +1,370 @@
+        predicted_scores = np.zeros(len(data))
+        for query in query_indexes:
+            results = np.zeros(len(query_indexes[query]))
+            for tree in self.trees:
+                results += self.learning_rate * tree.predict(data[query_indexes[query], 1:])
+            predicted_scores[query_indexes[query]] = results
+        return predicted_scores
+
+    def _validate(self, data, k=10):
+        data = np.array(data)
+        query_indexes = self._group_queries(data, 1)
+        average_ndcg = []
+        predicted_scores = np.zeros(len(data))
+        for query in query_indexes:
+            results = np.zeros(len(query_indexes[query]))
+            for tree in self.trees:
+                results += self.learning_rate * tree.predict(data[query_indexes[query], 2:])
+            predicted_sorted_indexes = np.argsort(results)[::-1]
+            t_results = data[query_indexes[query], 0]
+            t_results = t_results[predicted_sorted_indexes]
+            predicted_scores[query_indexes[query]] = results
+            ndcg_val = ndcg_at_k(r=t_results, k=10)
+            self.logger.info('Qid: {}, ndcg {} '.format(query, ndcg_val))
+            average_ndcg.append(ndcg_val)
+        average_ndcg = np.nanmean(average_ndcg)
+        return average_ndcg, predicted_scores
+
+    def set_tunable_parameters(self, min_samples_split, max_depth, min_samples_leaf, max_leaf_nodes, number_of_trees=5,
+                               learning_rate=1e-4, **kwargs):
+        """
+            Set the tunable hyperparameters of the DecisionTree model used in LambdaMART
+
+            Parameters
+            ----------
+            min_samples_split : int
+                Number of samples required to split the internal node
+            max_depth : int
+                Maximum depth of the tree
+            min_samples_leaf : int
+                Minimum number of samples required to be at the leaf node
+            max_leaf_nodes : int
+                These are the maximum number of leaf nodes used to grow the tree
+            number_of_trees : int
+                The maximum number of trees that are to be trained for the ensemble.
+            learning_rate : float
+                learning rate for the LambdaMART algorithm
+        """
+        self.min_samples_split = min_samples_split
+        self.max_depth = max_depth
+        self.min_samples_leaf = min_samples_leaf
+        self.max_leaf_nodes = max_leaf_nodes
+        self.number_of_trees = number_of_trees
+        self.learning_rate = learning_rate
+
+
+def compute_lambda_weights(args):
+    """
+        This is used by the LambdaMART learner. Computes the lambda targets that are considered to be the relationship between the different features and the ranking of a document. These lambdas are the learning target for the base learner.
+        The parameters to this function are passed in the form of a zipped iterator.
+
+        Parameters
```

Comment:
Those are not the parameters of the function. It only receives args (which is not good).